src/libcore/num/f32.rs

```diff
@@ -474,4 +475,85 @@ impl f32 {
         unsafe { mem::transmute(v) }
     }
+
+    /// A 3-way version of the IEEE `totalOrder` predicate.
+    ///
+    /// Because of negative zero and NaNs, the ordinary comparison operators
+    /// for `f32` do not represent a total order.  This method, however,
+    /// defines an ordering between all distinct bit patterns:
+    ///
+    /// - For normal numbers, the ordering matches the comparison operators.
+    /// - Zeros are between the positive and negative normal numbers, with
+    ///   the positive zero greater than the negative zero.
+    /// - Positive infinity is greater than all normal numbers; negative
+    ///   infinity is less than all normal numbers.
+    /// - Positive NaNs are greater than all other values, ordered amongst
+    ///   themselves by their payloads; Negative NaNs are less than all other
+    ///   values, ordered amongst in reverse order of their payloads.
+    ///
+    /// # Examples
+    ///
+    /// Normal numbers:
+    /// ```
+    /// #![feature(float_total_cmp)]
+    /// use std::cmp::Ordering::*;
+    ///
+    /// assert_eq!(f32::total_cmp(1.0, 2.0), Less);
+    /// assert_eq!(f32::total_cmp(1.0, 1.0), Equal);
+    /// assert_eq!(f32::total_cmp(2.0, 1.0), Greater);
+    /// assert_eq!(f32::total_cmp(-1.0, 1.0), Less);
+    /// assert_eq!(f32::total_cmp(1.0, -1.0), Greater);
+    /// assert_eq!(f32::total_cmp(-1.0, -2.0), Greater);
+    /// assert_eq!(f32::total_cmp(-1.0, -1.0), Equal);
+    /// assert_eq!(f32::total_cmp(-2.0, -1.0), Less);
+    /// ```
+    ///
+    /// Zeros, infinities, and NaNs:
+    /// ```
+    /// #![feature(float_total_cmp)]
+    /// use std::cmp::Ordering::*;
+    ///
+    /// assert_eq!(f32::partial_cmp(&0.0, &-0.0), Some(Equal));
+    /// assert_eq!(f32::total_cmp(-0.0, 0.0), Less);
+    /// assert_eq!(f32::total_cmp(0.0, -0.0), Greater);
+    ///
+    /// assert_eq!(f32::partial_cmp(&std::f32::NAN, &std::f32::NAN), None);
+    /// assert_eq!(f32::total_cmp(-std::f32::NAN, std::f32::NAN), Less);
+    /// assert_eq!(f32::total_cmp(std::f32::NAN, std::f32::NAN), Equal);
+    /// assert_eq!(f32::total_cmp(std::f32::NAN, -std::f32::NAN), Greater);
+    /// assert_eq!(f32::total_cmp(std::f32::NAN, std::f32::INFINITY), Greater);
+    /// assert_eq!(f32::total_cmp(-std::f32::NAN, -std::f32::INFINITY), Less);
+    /// ```
+    ///
+    /// Using this to sort floats:
+    /// ```
+    /// #![feature(float_total_cmp)]
+    ///
+    /// let mut a = [
+    ///     1.0,
+    ///     -1.0,
+    ///     0.0,
+    ///     -0.0,
+    ///     std::f32::NAN,
+    ///     -std::f32::NAN,
+    ///     std::f32::INFINITY,
+    ///     std::f32::NEG_INFINITY,
+    /// ];
+    /// a.sort_by(|&x, &y| f32::total_cmp(x, y));
+    /// assert_eq!(
+    ///     format!("{:?}", a),
+    ///     "[NaN, -inf, -1.0, -0.0, 0.0, 1.0, inf, NaN]"
+    /// );
+    /// ```
+    #[unstable(feature = "float_total_cmp", issue = "88888888")]
+    #[inline]
+    pub fn total_cmp(self, other: f32) -> cmp::Ordering {
```

Comment:
Could this function take references like `Ord::cmp` does? It would allow writing things like `a.sort_by(f32::total_cmp)`.

Reply:
I contemplated that, but decided that would be inconsistent with all the other things that would take references if f32 wasn't copy, but takes owned because it is copy.  I note that there's currently not a single `&self` or `&f32` in the inherent impl methods today.

Hopefully the copy type ergonomics stuff and better fn type coercions will make `.sort_by(f32::total_cmp)` _just work_ sometime in the future.

Reply:
I was wondering about this too. If one just can't `.sort_by(f32::total_cmp)` people will definitely reach to unsafe code for going from `&[f32]`s to `&[Total<f32>]` and vice-versa.

Reply:
Well, the other solution to that would be to just [make it safe.](https://github.com/rust-lang/rust/issues/49792#issuecomment-379638786)..